cabal-install/Distribution/Client/Tar.hs

```diff
@@ -453,12 +461,28 @@ mapEntries :: (Entry -> Entry) -> Entries -> Entries
 mapEntries f = foldrEntries (Next . f) Done Fail
 
 filterEntries :: (Entry -> Bool) -> Entries -> Entries
-filterEntries p =
-  foldrEntries
-    (\entry rest -> if p entry
-                      then Next entry rest
-                      else rest)
-    Done Fail
+filterEntries p = isoR . filterEntriesW (return . p)
```

Comment:
Do `filterEntriesW` and `foldrEntriesW` have to use `WriterT` or can we generalise them to an arbitrary `Monad`? I don't see anything `WriterT`-specific in the implementation.